src/Common/Core/Test/Services/ServiceManagerTests.cs

```diff
@@ -42,26 +36,74 @@ public class ServiceManagerTests {
         }
 
         [Test]
-        public void DoubleAdd() {
-            _serviceManager.AddService<I1>(new C1());
-            
-            Action a = () => _serviceManager.AddService<I1>(new C1());
-            a.ShouldThrow<InvalidOperationException>();
-        }
-
-        [Test]
         public void AddRemove() {
             var s = new C1();
             _serviceManager.AddService(s);
             _serviceManager.RemoveService(s);
             _serviceManager.GetService<I1>().Should().BeNull();
 
-            _serviceManager.AddService(s as I1);
+            _serviceManager.AddService((I1) s);
             _serviceManager.RemoveService<I1>();
             _serviceManager.GetService<I1>().Should().BeNull();
         }
 
         [Test]
+        public void AddRemoveDerived() {
+            var d = new Derived(_serviceManager);
+            _serviceManager.GetService<Derived>().Should().NotBeNull();
+            _serviceManager.GetService<Base>().Should().NotBeNull();
+            _serviceManager.GetService<I1>().Should().NotBeNull();
+            _serviceManager.GetService<I2>().Should().NotBeNull();
+
+            d.Dispose();
+            _serviceManager.GetService<Derived>().Should().BeNull();
+            _serviceManager.GetService<Base>().Should().BeNull();
+            _serviceManager.GetService<I1>().Should().BeNull();
+            _serviceManager.GetService<I2>().Should().BeNull();
+        }
+
+        [Test]
+        public void AddRemoveLazy01() {
+            _serviceManager.AddService<I2, Derived>();
+            _serviceManager.RemoveService<I2>();
+
+            _serviceManager.GetService<I1>().Should().BeNull();
+            _serviceManager.GetService<I2>().Should().BeNull();
+        }
+
+        [Test]
+        public void AddRemoveLazy02() {
+            _serviceManager.AddService<I2, Derived>();
+            _serviceManager.RemoveService<I1>();
+
+            _serviceManager.GetService<I1>().Should().BeNull();
+            _serviceManager.GetService<I2>().Should().BeNull();
+        }
+
+        [Test]
+        public void AddRemoveLazy03() {
+            _serviceManager.AddService<I2, Derived>();
+            _serviceManager.RemoveService<Base>();
```

Comment:
To the above - for lazy case there is `AddService<I1, A1>` (if someone looks for `I1`, instantiate `A1`) and generally `AddService<A>` is not very useful.